src/controller/proxy/controller.go

```diff
@@ -175,35 +213,64 @@ func (c *controller) UseLocalManifest(ctx context.Context, art lib.ArtifactInfo,
+		mediaType, payload, err := man.Payload()
+		if err != nil {
+			log.Errorf("Failed to get payload for manifest with digest: %v, error: %v", dig, err)
+		}
+		return true, &Manifest{content: payload, digest: dig, contentType: mediaType}, nil
+	}
+	log.Errorf("Failed to get manifest from local repo, error: %v", err)
+
 	var content []byte
 	var contentType string
 	if c.cache == nil {
 		return a != nil && string(desc.Digest) == a.Digest, nil, nil // digest matches
 	}
-	// Pass digest to the cache key, digest is more stable than tag, because tag could be updated
-	if len(art.Digest) == 0 {
-		art.Digest = string(desc.Digest)
-	}
+
+	// attempt to retrieve manifest list from cache
 	err = c.cache.Fetch(ctx, manifestListKey(art.Repository, art), &content)
-	if err != nil {
-		if errors.Is(err, cache.ErrNotFound) {
-			log.Debugf("Digest is not found in manifest list cache, key=cache:%v", manifestListKey(art.Repository, art))
-		} else {
-			log.Errorf("Failed to get manifest list from cache, error: %v", err)
+	if err == nil {
+		// manifest list was found in cache
+		err = c.cache.Fetch(ctx, manifestListContentTypeKey(art.Repository, art), &contentType)
+		if err != nil {
+			log.Debugf("failed to get the manifest list content type, not use local. error:%v", err)
+			return false, nil, nil
 		}
-		return a != nil && string(desc.Digest) == a.Digest, nil, nil
+		log.Debugf("Get the manifest list with key=cache:%v", manifestListKey(art.Repository, art))
+		return true, &ManifestList{content, string(desc.Digest), contentType}, nil
 	}
-	err = c.cache.Fetch(ctx, manifestListContentTypeKey(art.Repository, art), &contentType)
-	if err != nil {
-		log.Debugf("failed to get the manifest list content type, not use local. error:%v", err)
-		return false, nil, nil
+	if errors.Is(err, cache.ErrNotFound) {
+		log.Debugf("Digest is not found in manifest list cache, key=cache:%v", manifestListKey(art.Repository, art))
+	} else {
+		log.Errorf("Failed to get manifest list from cache, error: %v", err)
 	}
-	log.Debugf("Get the manifest list with key=cache:%v", manifestListKey(art.Repository, art))
-	return true, &ManifestList{content, string(desc.Digest), contentType}, nil
+
+	// neither manifest was found in local repo nor manifest list was found in cache
+	return a != nil && string(desc.Digest) == a.Digest, nil, nil
+}
+
+func manifestKey(repo string, art lib.ArtifactInfo) string {
+	// actual redis key format is cache:manifest:<repo name>:<tag> or cache:manifest:<repo name>:sha256:xxxx
+	return "manifest:" + repo + ":" + getReference(art)
+}
+
+func manifestContentTypeKey(rep string, art lib.ArtifactInfo) string {
+	return manifestKey(rep, art) + ":contenttype"
 }
 
 func manifestListKey(repo string, art lib.ArtifactInfo) string {
-	// actual redis key format is cache:manifestlist:<repo name>:<tag> or cache:manifestlist:<repo name>:sha256:xxxx
+	// actual redis key format is cache:manifest:<repo name>:<tag> or cache:manifest:<repo name>:sha256:xxxx
```

Comment:
The manifestListKey is only cache the manfest List, not used to cache a normal manifest

Reply:
Sorry, updating this comment was a mistake. I will revert it.